server.go

```diff
@@ -295,26 +303,76 @@ type Server struct {
 	DecorateReader DecorateReader
 	// DecorateWriter is optional, allows customization of the process that writes raw DNS messages.
 	DecorateWriter DecorateWriter
-
+	// UDP packet or TCP connection queue
+	queue chan *response
+	// Workers count
+	workersCount int32
 	// Shutdown handling
 	lock    sync.RWMutex
 	started bool
 }
 
+func (srv *Server) worker(w *response) {
+	workersCount := atomic.LoadInt32(&srv.workersCount)
+	if workersCount > maxWorkersCount {
+		w.Close()
+		return
+	}
+	atomic.AddInt32(&srv.workersCount, 1)
+
+	srv.serve(w)
+
+	count := 0
+	timeout := time.After(idleWorkerTimeout)
+LOOP:
+	for {
+		select {
+		case w, ok := <-srv.queue:
+			if !ok {
+				break LOOP
+			}
+			count += 1
+			srv.serve(w)
+		case <-timeout:
+			if count == 0 {
+				break LOOP
+			}
+			count = 0
+			timeout = time.After(idleWorkerTimeout)
+		}
+	}
+	atomic.AddInt32(&srv.workersCount, -1)
+}
+
+func (srv *Server) spawnWorker(w *response) {
+	select {
+	case srv.queue <- w:
+	default:
+		go srv.worker(w)
+	}
+}
+
 // ListenAndServe starts a nameserver on the configured address in *Server.
 func (srv *Server) ListenAndServe() error {
 	srv.lock.Lock()
 	defer srv.lock.Unlock()
 	if srv.started {
 		return &Error{err: "server already started"}
 	}
+
+	if srv.Handler == nil {
+		srv.Handler = DefaultServeMux
```

Comment:
I’d move this conditional into `srv.serveDNS`, there’s no need to modify `srv` here or below.

Reply:
It’s not wrong by any means, but it’s very much *not* idiomatic. Go code like this very rarely mutates public fields. (This is the `net/http` case: https://golang.org/src/net/http/server.go?s=82393:82459#L2676).

Also a single if statement will have zero effect on performance.

(You don’t need to pass it in as an argument like it was before, but you can. It should be just put in `serveDNS`).

Reply:
Done